src/net64/emulator/m64plus.cpp

```diff
@@ -104,73 +92,126 @@ void Mupen64Plus::unload_rom()
+        {
+            switch (new_value)
+            {
+            case M64PTypes::M64EMU_STOPPED:
+                // This is to guarantee execute() owns mutex_ when do_cmd returns
+                mutex_.lock();
+                state_ = State::Stopped;
+                break;
+            case M64PTypes::M64EMU_RUNNING:
+                if(state_ == State::Starting)
+                {
+                    // To guarantee that stop() cannot be entered while starting we kept mutex_
+                    // now that we are running we want to be able to stop, so unlock it.
+                    state_ = State::Running;
+                    mutex_.unlock();
+                    emulation_started();
+                }
+                break;
+            case M64PTypes::M64EMU_PAUSED:
+                mutex_.lock();
+                state_ = State::Paused;
+                mutex_.unlock();
+                break;
+            default:
+                assert(false);
+            }
+        }
+    });
+    loginfo_noexcept("Starting n64 emulation");
     auto ret{core_.do_cmd(M64PTypes::M64CMD_EXECUTE, 0, nullptr)};
     detach_plugins();
-    running_ = false;
+    assert(state_ == State::Stopped);
 
     if(failed(ret))
     {
         std::system_error err{make_error_code(ret), "Failed to execute ROM image"};
+        mutex_.unlock();
         logger()->error(err.what());
         throw err;
     }
 
-    logger()->info("Stopped n64 emulation");
+    loginfo_noexcept("Stopped n64 emulation");
+    mutex_.unlock();
 }
 
 void Mupen64Plus::stop()
 {
-    using namespace M64PTypes;
+    // Take ownership of mutex_, this guarantees nobody will change state_
+    // while we own mutex_.
+    mutex_.lock();
 
-    if(!running_)
+    // Do nothing if already stopped or stopping
+    if(state_ == State::Stopped || state_ == State::Stopping)
+    {
+        mutex_.unlock();
         return;
+    }
 
-    // Wait for emulator to be fully started
-    m64p_emu_state state{};
-    do
-    {
-        core_.do_cmd(M64CMD_CORE_STATE_QUERY, M64CORE_EMU_STATE, &state);
-    }while(state != M64EMU_RUNNING);
+    // We cannot be starting because execute() owns mutex_ until state_ == State::Started,
+    // because we own mutex_ execute() cannot be between Starting and Started.
+    assert(state_ != State::Starting);
 
-    // Wait for emulator to be fully stopped
-    do
-    {
-        core_.do_cmd(M64CMD_STOP, 0, nullptr);
-        core_.do_cmd(M64CMD_CORE_STATE_QUERY, M64CORE_EMU_STATE, &state);
-    }while(state != M64EMU_STOPPED);
-}
+    // Stop the emulator
+    loginfo_noexcept("Stopping n64 emulation");
+    core_.do_cmd(M64PTypes::M64CMD_STOP, 0, nullptr);
+    state_ = State::Stopping;
+    mutex_.unlock();
 
-Mupen64Plus::Core& Mupen64Plus::core()
-{
-    return core_;
+    // Wait for the emulator to be stopped
+    while(state_ != State::Stopped)
+    {
+        core_.do_cmd(M64PTypes::M64CMD_STOP, 0, nullptr);
```

Comment:
Stop is implemented like this:

```
void stop_device(struct device* dev)
{
    /* set stop flag so that r4300 execution will be stopped at next interrupt */
    *r4300_stop(&dev->r4300) = 1;
}
```

However during startup:

```
/* Startup message on the OSD */
    osd_new_message(OSD_MIDDLE_CENTER, "Mupen64Plus Started...");

    g_EmulatorRunning = 1;
    StateChanged(M64CORE_EMU_STATE, M64EMU_RUNNING); //here

    poweron_device(&g_dev);
    pif_bootrom_hle_execute(&g_dev.r4300);
    run_device(&g_dev);
```

We receive M64EMU_RUNNING before `poweron_device`, inside of which there is a memset on the whole `r4300` struct. Anyway, the answer is basically: because of a bug in mupen. I'll open a PR that fixes this in core, but for now this workaround has to stay.

Reply:
I opened https://github.com/mupen64plus/mupen64plus-core/issues/681 so they will maybe fix it in core and also added `std::this_thread::yield()` in the loop to make it not so hot.